ocaml/xapi/xapi_pci_helpers.ml

```diff
@@ -41,6 +41,53 @@ let get_driver_name address =
       debug "get_driver_name: for %s failed with %s" address (Printexc.to_string e); 
       None
 
+let address_of_dev x =
+  let open Pci.Pci_dev in
+  Printf.sprintf "%04x:%02x:%02x.%d" x.domain x.bus x.dev x.func
+
+(* Check for a PCI device whether it is virtual and remember the result
+ * such that it can be looked up later *)
+module PCIcache: sig
+  type t
+  type addr = string (* "0000:37:00.4" *)
+  val make: unit -> t
+  val is_virtual: t -> addr -> bool
+end = struct
+  type t = (string, bool) Hashtbl.t
+  type addr = string
+
+  (** [is_virtual_pci "0000:37:00.4"] is true, if this designates a
+   * virtual PCI function (VF), false otherwise. Only a VF has a "physfn"
+   * symbolic link.
+   *)
+  let is_virtual addr =
+    let path = Printf.sprintf "/sys/bus/pci/devices/%s/physfn" addr in
+    try
+      ignore @@ Unix.readlink path; true
```

Comment:
Would Sys.file_exists work instead?

Reply:
This is almost equivalent. `readlink` would fail if this is not a symbolic link, whereas `Sys.file_exsts` would succeed. Neither would fail if the target of the symbolic link does not exist.